src/librustc_mir/transform/check_consts/resolver.rs

```diff
@@ -0,0 +1,327 @@
+use rustc::mir::visit::Visitor;
+use rustc::mir::{self, BasicBlock, Local, Location};
+use rustc_data_structures::bit_set::BitSet;
+
+use std::cell::RefCell;
+use std::marker::PhantomData;
+use std::rc::Rc;
+
+use crate::dataflow::{self as old_dataflow, generic as dataflow};
+use super::{Item, Qualif};
+use self::old_dataflow::IndirectlyMutableLocals;
+
+/// A `Visitor` that propagates qualifs between locals. This defines the transfer function of
+/// `FlowSensitiveAnalysis` as well as the logic underlying `TempPromotionResolver`.
+///
+/// This transfer does nothing when encountering an indirect assignment. Consumers should rely on
+/// the `IndirectlyMutableLocals` dataflow pass to see if a `Local` may have become qualified via
+/// an indirect assignment or function call.
+struct TransferFunction<'a, 'mir, 'tcx, Q> {
+    item: &'a Item<'mir, 'tcx>,
+    qualifs_per_local: &'a mut BitSet<Local>,
+
+    _qualif: PhantomData<Q>,
+}
+
+impl<Q> TransferFunction<'a, 'mir, 'tcx, Q>
+where
+    Q: Qualif,
+{
+    fn new(
+        item: &'a Item<'mir, 'tcx>,
+        qualifs_per_local: &'a mut BitSet<Local>,
+    ) -> Self {
+        TransferFunction {
+            item,
+            qualifs_per_local,
+            _qualif: PhantomData,
+        }
+    }
+
+    fn initialize_state(&mut self) {
+        self.qualifs_per_local.clear();
+
+        for arg in self.item.body.args_iter() {
+            let arg_ty = self.item.body.local_decls[arg].ty;
+            if Q::in_any_value_of_ty(self.item, arg_ty).unwrap() {
+                self.qualifs_per_local.insert(arg);
+            }
+        }
+    }
+
+    fn assign_qualif_direct(&mut self, place: &mir::Place<'tcx>, value: bool) {
+        debug_assert!(!place.is_indirect());
+
+        match (value, place) {
+            (true, mir::Place { base: mir::PlaceBase::Local(local), .. }) => {
+                self.qualifs_per_local.insert(*local);
+            }
+
+            // For now, we do not clear the qualif if a local is overwritten in full by
+            // an unqualified rvalue (e.g. `y = 5`). This is to be consistent
+            // with aggregates where we overwrite all fields with assignments, which would not
+            // get this feature.
```

Comment:
For background, the goal is to minimize confusion when you change a variable in a `const` to a tuple.

```rust
{
    let mut x = Some(NeedsDrop);
    /* ... */
    x = None;
    // Would work if qualifs were overwritten
}

{
    let mut x = (Some(NeedsDrop), true);
    /* ... */
    x = (None, false);
    // Would break if x is overwritten field-by-field
}
```